### PR TITLE
Provide range-based numeric generators and javax.time generators #530

### PR DESCRIPTION
**Pull request by sksamuel**

### CODE REVIEW

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 1]
```diff
@@ -363,6 +380,33 @@ interface Gen<T> {
       override fun random(): Sequence<UUID> = generateSequence { UUID.randomUUID() }
     }
 
+    fun localDate(minYear: Int = 1, maxYear: Int = 2030): Gen<LocalDate> = object : Gen<LocalDate> {
+      override fun constants(): Iterable<LocalDate> = emptyList()
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: Interesting dates to be constants: February 28th, February 29th, Jan 1st and Dec 31st

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 2]
```diff
@@ -363,6 +380,33 @@ interface Gen<T> {
...
+      override fun random(): Sequence<LocalDate> = generateSequence { LocalDate.of(RANDOM.nextInt(maxYear - minYear) + minYear, RANDOM.nextInt(12) + 1, RANDOM.nextInt(31) + 1) }
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: Bad randoms. Will generate February 30th, April 31st

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 3]
```diff
@@ -363,6 +380,33 @@ interface Gen<T> {
       override fun random(): Sequence<UUID> = generateSequence { UUID.randomUUID() }
     }
 
+    fun localDate(minYear: Int = 1, maxYear: Int = 2030): Gen<LocalDate> = object : Gen<LocalDate> {
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: I think that minYear should default to 1970, but I liked to be able to customize!

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 4]
```diff
@@ -363,6 +380,33 @@ interface Gen<T> {
...
+      override fun constants(): Iterable<LocalDate> = emptyList()
+      override fun random(): Sequence<LocalDate> = generateSequence { LocalDate.of(RANDOM.nextInt(maxYear - minYear) + minYear, RANDOM.nextInt(12) + 1, RANDOM.nextInt(31) + 1) }
+    }
+
+    fun localTime(): Gen<LocalTime> = object : Gen<LocalTime> {
+      override fun constants(): Iterable<LocalTime> = emptyList()
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: Maybe 00:00 and 23:59 could be good constants here?

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 5]
```diff
@@ -363,6 +380,33 @@ interface Gen<T> {
...
+      override fun random(): Sequence<LocalTime> = generateSequence { LocalTime.of(RANDOM.nextInt(24), RANDOM.nextInt(60), RANDOM.nextInt(60)) }
+    }
+
+    fun localDateTime(minYear: Int = 1, maxYear: Int = 2030): Gen<LocalDateTime> = object : Gen<LocalDateTime> {
+      override fun constants(): Iterable<LocalDateTime> = emptyList()
+      override fun random(): Sequence<LocalDateTime> = generateSequence { LocalDateTime.of(RANDOM.nextInt(maxYear - minYear) + minYear, RANDOM.nextInt(12) + 1, RANDOM.nextInt(31) + 1, RANDOM.nextInt(24), RANDOM.nextInt(60), RANDOM.nextInt(60)) }
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: Maybe better to use the other generators here. `generatedLocalDate.atTime(generatedTime)`

Comment (LeoColman) [reply kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: Prevents extra maintance points (although unlikely)

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 6]
```diff
@@ -363,6 +380,33 @@ interface Gen<T> {
...
+    }
+
+    fun duration(): Gen<Duration> = object : Gen<Duration> {
+      private fun randomDuration(): Duration = when (RANDOM.nextInt(5)) {
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: We can reduce this code a bit, and include more constants:

```kotlin
val wantedUnits = ChronoUnit.values().filter { it <= ChronoUnit.WEEKS } // Will get all crono units that represent a week or less

return Duration.of(Random.nextLong(), wantedUnits.random())
```

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 7]
```diff
@@ -374,6 +418,26 @@ interface Gen<T> {
       override fun shrinker(): Shrinker<Double>? = DoubleShrinker
     }
 
+    /**
+     * Returns a [[Gen]] which is the same as [[Gen.double]]] but does not include infinity or NaN.
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: Does not include -INFINITY, +INFINITY or NaN

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 8]
```diff
@@ -374,6 +418,26 @@ interface Gen<T> {
...
+     */
+    fun numericDoubles(): Gen<Double> = object : Gen<Double> {
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: For this to be range-based, we should add a `minValue` and `maxValue`

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 9]
```diff
@@ -374,6 +418,26 @@ interface Gen<T> {
...
+      val literals = listOf(0.0, 1.0, -1.0, 1e300, Double.MIN_VALUE, Double.MAX_VALUE)
+      override fun constants(): Iterable<Double> = literals
+      override fun random(): Sequence<Double> = generateSequence { RANDOM.nextDouble() }
+      override fun shrinker(): Shrinker<Double>? = DoubleShrinker
+    }
+
+    /**
+     * Returns a [[Gen]] which is the same as [[Gen.float]]] but does not include infinity or NaN.
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: Same comments from Double Gen

**Review by LeoColman — APPROVED** [review]

**Review by LeoColman — CHANGES_REQUESTED** [review]
Some changes should be made

**Comment by Kerooker** [comment]
Also missing tests for the new gens and functions.

**Comment by sksamuel** [comment]
Good feedback.

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt (hunk 6 above)
Comment (LeoColman) [reply kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt]: Would be also cool to have a Gen<> for Period

**Comment by LeoColman** [comment]
@sksamuel 
I tweaked some of the generators. Could you take a look, please?

**Comment by sksamuel** [comment]
Looks excellent. Feel free to merge.

**Pull request merged**